### ISSUE TITLE
Integrate navigation bars from different viewers?

### ISSUE THREAD

**Issue by uliska** [wbsoft/frescobaldi#710]
I add this to the "Manuscript" milestone although it's explicitly not meant as something that has to be done before the initial release.

The music view has its navigational controls in the toolbar of the main window, now we're adding a navigation bar at the top but *inside* the manuscript viewer. Furthermore the toolbars share *a lot* of common options. So it would be a good idea to discuss how we can integrate them - also from the perspective that there is at least one more (current) viewer available and that there might come some more in the future.

There are two aspects to that: the GUI layout and the coding.

@wbsoft  has suggested to some day move the music view controls to the top of the panel too. But OTOH I think having *all* toolbars in the global toolbar may be valuable with regard to screen estate. If I have the music view and the manuscript viewer on top of each other two toolbars would hurt quite a bit.

I have two ideas how the toolbars could be integrated, at the same time layout- and implementation-wise.

a)
Create an abstract base class and have the toolbars for the different viewers inherit from this.

b)
Use only *one* navigational toolbar for all viewers and let it switch "focus". The switch can be triggered either by the viewer getting focus or by a "literal" switch control on the toolbar.  The action of switching context for the toolbar would have to hide/show/update necessary controls and reconnect the signals to the slots of the appropriate viewer.
It might be a little complicated to get this right so the user gets a smooth workflow and never suffers from ambiguitites. But it might be worth it for the saved screen estate and the consistency of the interface.

Common controls are:

* Zoom: +/- and the combobox
* Pager: </> and the pagenum control
* Print music (I'm not sure if it makes sense to print the document from the manuscript viewer)?

Currently the document chooser is only useful for the music view, but when the manuscript viewer will support multiple manuscripts they should be accessible from the same control.

The help button is currently only available for the manuscript viewer but could also be good to have for any viewer.

The Open/Close controls are (and probably remain) exclusive for the manuscript viewer. But they might be changed to smaller image buttons and/or menu access some day.

The manuscript viewer will probably need more controls when it gets more mature. But I think these could well go into a menu/contextmenu interface.

**Comment by PeterBjuhr** [comment]
I think it could be good to settle this before we go any further with the controls 

>I have two ideas how the toolbars could be integrated, at the same time layout- and implementation-wise.

>a) Create an abstract base class and have the toolbars for the different viewers inherit from this.

This is more or less already done. The icons are set. All that is left (if I haven't missed anything) is adding the actions to the toolbar. 

>b) Use only one navigational toolbar for all viewers and let it switch "focus". The switch can be triggered either by the viewer getting focus or by a "literal" switch control on the toolbar. The action of switching context for the toolbar would have to hide/show/update necessary controls and reconnect the signals to the slots of the appropriate viewer.
It might be a little complicated to get this right so the user gets a smooth workflow and never suffers from ambiguitites. But it might be worth it for the saved screen estate and the consistency of the interface.

A switch sounds much more inconvenient than two different toolbars to my mind. And being dependent of focus to decide which viewer to control doesn't sound like a userfriendly solution either; you basically then always have to make sure that the right viewer has focus before being able to use the controls.

**Comment by uliska** [comment]
OK, then let's go for the first approach.

**Comment by uliska** [comment]
OK, I've added the zoomer and the pager actions.

I also wanted to move the open/close buttons inside the new toolbar (so one could later handle them as one) but that didn't seem to work.
I don't think this requires the issue to be kept open, though.

**Comment by PeterBjuhr** [comment]
>I also wanted to move the open/close buttons inside the new toolbar (so one could later handle them as one) but that didn't seem to work.

They can be added with `addWidget` (not `addAction` as the others).

**Comment by PeterBjuhr** [comment]
But maybe it's better to redesign them as actions aswell...

**Comment by PeterBjuhr** [comment]
Should we also use icons instead of buttons?

**Comment by uliska** [comment]
Yes, that would be more consistent and would save space.

**Comment by uliska** [comment]
I have integrated them in as widgets in the toolbar for now (45c6eb8c)

**Comment by PeterBjuhr** [comment]
Compared to the music view toolbar we are still missing the document chooser and the print button. At least in my setup there's room to add those aswell. But in the future other button may compete for the space.

@uliska with your approval I'll add those two items too!?

**Comment by uliska** [comment]
I stopped adding the print button because I ran into a missing action.
But that shouldn't be that hard to implement, so go ahead.

With regard to the document chooser I'd say we *will* need it, but for
now we don't have the means to populate it with alternatives. In the
music view it is filled with the lilypond-generated files. In our case
we *could* change the file open dialog to allow multiple files again and
place these in the document chooser. In that case we'd have to update
the session logic too.

But on the long run I think we will need a way to tie manuscripts to
input documents. And I'm still sure this will be a metadata file that
lies beside the input file, say master-document.manuscripts beside
master-document.ly. So far it doesn't matter if that will be an MEI file
or a simpler (e.g. JSON) structure, but I think that's the way to go.

OTOH we could *still* have the option to open multiple files through the
dialog when such a metadata file isn't present.

**Comment by PeterBjuhr** [comment]
>But on the long run I think we will need a way to tie manuscripts to
input documents. 

For my part I'm very happy with working on a session basis. I actually have real life case I'm working with at the moment where I use the manuscript viewer (have I mentioned this before?). Which is very convenient for testing purposes. I would definitely say that my basic needs pretty soon is fulfilled with the current implementation.

>And I'm still sure this will be a metadata file that
lies beside the input file, say master-document.manuscripts beside
master-document.ly. So far it doesn't matter if that will be an MEI file
or a simpler (e.g. JSON) structure, but I think that's the way to go.

I'm of course interested in your *advanced* plans for the viewer, but I'm reluctant to start a discussion about it at the moment. Perhaps it's better to wait a little while and focus on the "release" (or we could in fact start a discussion labelled issue about it...).

>OTOH we could *still* have the option to open multiple files through the
dialog when such a metadata file isn't present.

So we can implement this for now!?

**Comment by uliska** [comment]
Am 15.07.2015 um 11:14 schrieb Peter Bjuhr:
>     But on the long run I think we will need a way to tie manuscripts to
>     input documents.
> 
> For my part I'm very happy with working on a session basis. I actually
> have real life case I'm working with at the moment where I use the
> manuscript viewer (have I mentioned this before?). 

No.

> Which is very
> convenient for testing purposes. I would definitely say that my basic
> needs pretty soon is fulfilled with the current implementation.

And would you benefit from having multiple manuscripts available in the
document chooser?

> 
>     And I'm still sure this will be a metadata file that
>     lies beside the input file, say master-document.manuscripts beside
>     master-document.ly. So far it doesn't matter if that will be an MEI file
>     or a simpler (e.g. JSON) structure, but I think that's the way to go.
> 
> I'm of course interested in your /advanced/ plans for the viewer, but
> I'm reluctant to start a discussion about it at the moment. Perhaps it's
> better to wait a little while and focus on the "release" (or we could in
> fact start a discussion labelled issue about it...).

Yes, that was what I intended.

> 
>     OTOH we could /still/ have the option to open multiple files through the
>     dialog when such a metadata file isn't present.
> 
> So we can implement this for now!?

Yes, that should be possible.

**Comment by PeterBjuhr** [comment]
>And would you benefit from having multiple manuscripts available in the
document chooser?

I haven't thought about, I have been mainly focused on the consistency with the music viewer and with the fact that the logic for this is already (mostly) in place.

But I guess it could be useful. I have several "manuscript" pdf:s relating to the project so it could be good to have them all saved in the session for easy access.

On the other hand I can hardly have all LilyPond files open at the same time so I wouldn't consider it inconvenient having to open a document once in a while.

**Comment by uliska** [comment]
In any case, I don't think it would be a problem if multiple documents
can be open at once. So the steps seem to be:

* add the document chooser
* change the file open dialog so multiple files can be selected and fed
into the document chooser
* reconsider how that relates to the session handling (update save/load
session functions)

If you take care of the first two I can look into the third (as I've
already been looking at the sessions)

**Comment by PeterBjuhr** [comment]
Great!

**Comment by uliska** [comment]
Opened two new issues because this closed issue contained orphaned TODO items.
